pandas/core/groupby/groupby.py

```diff
@@ -1688,6 +1690,72 @@ def nth(self, n, dropna=None):
 
         return result
 
+    def quantile(self, q=0.5, interpolation='linear'):
+        """
+        Return group values at the given quantile, a la numpy.percentile.
+
+        Parameters
+        ----------
+        q : float or array-like, default 0.5 (50% quantile)
+            0 <= q <= 1, the quantile(s) to compute
+        interpolation : {'linear', 'lower', 'higher', 'midpoint', 'nearest'}
+            Method to use when the desired quantile falls between two points.
+
+        Returns
+        -------
+        Series or DataFrame
+            Return type determined by caller of GroupBy object.
+
+        See Also
+        --------
+        Series.quantile : Similar method for Series
+        DataFrame.quantile : Similar method for DataFrame
+        numpy.percentile : NumPy method to compute qth percentile
+
+        Examples
+        --------
+        >>> df = pd.DataFrame(
+        ...    [['foo'] * 5 + ['bar'] * 5,
+        ...     [1, 2, 3, 4, 5, 5, 4, 3, 2, 1]],
+        ...    columns=['key', 'val'])
+        >>> df
+        """
+
+        inferences = {  # TODO (py27): replace with nonlocal
```

Comment:
this is pretty janky I would change to do this

```
def pre_processor(vals):
     ....
     return vals, inferences

def post_processor(vals, inferences):
     ....
```
basically returning the state (or an empty dict) in the pre and passing it to the post

Reply:
you will have to adjust the other callers of this but shouldn't be a big deal